Automattic/wp-calypso
# Lay foundations for function-based plan checks

Related to https://github.com/Automattic/wp-calypso/pull/23075 and 2-year plans in general (p9jf6J-eR-p2).

This PR in a more lengthy form and with more usage examples was here before: https://github.com/Automattic/wp-calypso/pull/23084

We are adding a set of plans to make 2-year subscriptions possible. Current codebase assumes that there is just one variant of each plan type (except for jetpack annual/monthly subscriptions). I can see at least the following problems off the top of my head:

1. Comparisons like `planSlug === PLAN_BUSINESS` would have to check also for `PLAN_BUSINESS_2_YEAR`
2. Code fragments like `<Banner plan={PLAN_BUSINESS} />` may need to receive `PLAN_BUSINESS_2_YEAR` if user is subscribed to a 2-year plan

This PR proposes a solution to above and similar issues. I went through all places where plan constants were used directly and sketched this code and it seems to cover all use cases I saw, namely:
* Get all business plans for both Jetpack and wp.com
* Get all premium plans of Jetpack
* Get personal plans of wp.com
* Get business plan for the same site (Jetpack/wp.com) as current user but for 2-year term
* Get premium plan for the same site (Jetpack/wp.com) as current user for the same term as current user
* Get the same type of plan as current user have, but with a 2-year term
* ...and more similar

Instead of using/comparing plans constants directly, we introduce some metadata to `PLANS_LIST` and also a few helper functions to encapsulate that logic.

We can get rid of some checks like `planSlug === PLAN_BUSINESS`  in favour of checking features, like here: https://github.com/Automattic/wp-calypso/pull/23076 - however this is not possible in all (most) cases.

With this PR, you can check like 
```js
planMatches(planSlug,  {type: TYPE_BUSINESS}); // or shortcut: isBusinessPlan( planSlug );
```

If you need to check just for jetpack or wp.com specific plans, you can do
```js
planMatches(planSlug, {type: TYPE_BUSINESS, group: GROUP_WPCOM})
```

Or, let's say you need a business plan with the same subscription term as active user and related to the same site (jetpack/wp) as current user's plan:
```js
findSimilarPlan(currentPlanSlug, {type: TYPE_BUSINESS})
```

Not only this would allow us to remove some boilerplate code from many components, but would also introduce a consistent way to work with plans - currently they are dealt with in a dozens of different ways.

Finally, with this PR we are able to get rid of other constants such as:
```js
export const JETPACK_PLANS = [
export const JETPACK_MONTHLY_PLANS = [
```
And have PLANS_LIST as a single source of truth allowing us to cross-reference anything in any way without other constants containing additional groupings.

**Test plan:**

1. Check if purchase/upgrade page with plans displays correct prices and looks properly.
1. Check if purchase/upgrade page with plans displays correct details about current plan and upgrade options for monthly users
